packages/bitcoin-bitcoinjs-lib-swap-provider/lib/BitcoinBitcoinJsLibSwapProvider.js

```diff
@@ -1,179 +1,229 @@
-    return this._redeemSwap(initiationTxHash, recipientAddress, refundAddress, expiration, true, secret)
+    const secretHash = bitcoin.crypto.sha256(Buffer.from(secret, 'hex')).toString('hex')
+    return this._redeemSwap(initiationTxHash, recipientAddress, refundAddress, expiration, true, secret, secretHash)
   }
 
   async refundSwap (initiationTxHash, recipientAddress, refundAddress, secretHash, expiration) {
     return this._redeemSwap(initiationTxHash, recipientAddress, refundAddress, expiration, false, undefined, secretHash)
   }
 
   async _redeemSwap (initiationTxHash, recipientAddress, refundAddress, expiration, isRedeem, secret, secretHash) {
+    const network = this._bitcoinJsNetwork
     const address = isRedeem ? recipientAddress : refundAddress
     const wif = await this.getMethod('dumpPrivKey')(address)
-    const wallet = bitcoin.ECPair.fromWIF(wif, this._bitcoinJsNetwork)
-    const script = this.createSwapScript(recipientAddress, refundAddress, secretHash || sha256(secret), expiration)
-    const scriptPubKey = padHexStart(script)
-    const p2shAddress = pubKeyToAddress(scriptPubKey, this._network.name, 'scriptHash')
-    const sendScript = this.getMethod('createScript')(p2shAddress)
+    const wallet = bitcoin.ECPair.fromWIF(wif, network)
+    const swapOutput = this.getSwapOutput(recipientAddress, refundAddress, secretHash, expiration)
+    const swapPaymentVariants = this.getSwapPaymentVariants(swapOutput)
 
     const initiationTxRaw = await this.getMethod('getRawTransactionByHash')(initiationTxHash)
     const initiationTx = await this.getMethod('decodeRawTransaction')(initiationTxRaw)
-    const voutIndex = initiationTx._raw.data.vout.findIndex((vout) => vout.scriptPubKey.hex === sendScript)
-    let vout = initiationTx._raw.data.vout[voutIndex]
-    const txfee = calculateFee(1, 1, 3)
 
-    vout.txid = initiationTxHash
-    vout.vSat = vout.value * 1e8
-    vout.script = Buffer.from(script, 'hex')
-    const walletRedeem = this.spendSwap(address, wallet, secret, isRedeem, txfee, vout, this._bitcoinJsNetwork, expiration)
-    return this.getMethod('sendRawTransaction')(walletRedeem)
-  }
+    let swapVout
+    let paymentVariantName
+    let paymentVariant
+    for (const voutIndex in initiationTx._raw.data.vout) {
+      const vout = initiationTx._raw.data.vout[voutIndex]
+      const paymentVariantEntry = Object.entries(swapPaymentVariants).find(([, payment]) => payment.output.toString('hex') === vout.scriptPubKey.hex)
+      if (paymentVariantEntry) {
+        paymentVariantName = paymentVariantEntry[0]
+        paymentVariant = paymentVariantEntry[1]
+        swapVout = vout
+      }
+    }
 
-  spendSwap (address, wallet, secret, isRedeem, txfee, vout, network, expiration) {
-    address = addressToString(address)
+    // TODO: Implement proper fee calculation that counts bytes in inputs and outputs
+    // TODO: use node's feePerByte
+    const txfee = calculateFee(1, 1, 3)
```

Comment:
Yes absolutely, i actually prefer to do a separate PR on the fee stuff which is why i left the comment. because generally the fee calculation in the lib as a whole is not correct. It should be based on the input and output sizes but we have hardcoded values there.